File: 프로그래머스/난이도별/Level3.네트워크/네트워크.py
# DFS 풀이
def solution(n, computers):
    answer = 0

    visited = [False for _ in range(n)]
    
    def DFS(x):
        visited[x] = True
        for i in range(n):
            if x != i and computers[x][i] == computers[i][x] == 1:
                if not visited[i]:
                    DFS(i)
            
    
    for com in range(n):
        if not visited[com]:
            DFS(com)
            answer += 1
            
    return answer
            
    


# BFS 풀이
from collections import deque

def solution(n, computers):
    answer = 0

    visited = [False for _ in range(n)]
    
    def BFS(com):
        queue = deque()
        queue.append(com)
        while queue:
            x = queue.popleft()
            visited[x] = True
            for i in range(n):
                if x != i and computers[x][i] == computers[i][x] == 1:
                    if not visited[i]:
                        queue.append(i)
            
    
    for com in range(n):
        if not visited[com]:
            BFS(com)
            answer += 1
            
    return answer
            
    

File: 프로그래머스/난이도별/Level3.네트워크/네트워크.js
function solution(n, computers) {
  let answer = 0;
  const visited = Array.from({ length: n }, () => false);

  function bfs(i) {
    const queue = [];
    queue.push(i);

    while (queue.length) {
      x = queue.shift();
      visited[x] = true;
      for (let j = 0; j < n; j++) {
        if (!visited[j] && computers[x][j] && computers[j][x] && x !== j) {
          queue.push(j);
        }
      }
    }
  }
  for (let i = 0; i < n; i++) {
    if (!visited[i]) {
      bfs(i);
      answer++;
    }
  }
  return answer;
}
